### DISCUSSION TITLE
Anisotropic Model: Lenght scale and Variance Specification

### DISCUSSION THREAD

**Discussion by chibueze-oguejiofor** [GeoStat-Framework/GSTools#178] [Q&A] [answered]
Hi All,

Having fitted two directional spherical variograms to my 2D dataset, i’ve obtained parameters for the models (two spherical models for the same data. One for the major direction, and the other for the transverse)
<img width="328" alt="Screen Shot 2021-06-07 at 11 30 17 PM" src="https://user-images.githubusercontent.com/30546741/121118755-670a4b00-c7e8-11eb-86db-504cfdee6279.png">
<img width="326" alt="Screen Shot 2021-06-07 at 11 30 02 PM" src="https://user-images.githubusercontent.com/30546741/121118753-6671b480-c7e8-11eb-86fd-6dddbd251816.png">

As you can see from the images above, the two lenght scales (range) in the main/transverse directions are roughly 90 and 180..also their sills are 1.89 and 3.4 respectively. I intend to define a single spherical anisotropic model that accounts for both length scales and sills (var), so as to generate some random fields. GStools allows for multiple length scales to be inputted as lists as shown below however, it doesn’t seem possible to input multiple sills(var)

```
model = gs.Spherical(dim=2, var=1.89, nugget=0.0, len_scale=[90.36, 180.62], angles=np.pi / 2)
```

My problem now is that: with my code above, i generate a model with two lenght scales (ranges) for the two transverse directions, BUT both lenght scales will have the same sill (var=1.89) as shown in the diagram below. How can i fix this (i.e input two different sills)?

<img width="332" alt="Screen Shot 2021-06-07 at 11 41 20 PM" src="https://user-images.githubusercontent.com/30546741/121119601-e8161200-c7e9-11eb-8eca-c682cbe102be.png">

I would greatly appreciate any help.

**Comment by MuellerSeb** [comment, marked as answer]
Hey there,

thanks for using GSTools.
Unfortunately, all variogram models in GSTools assume random fields that are stationary and ergodic.
That means, that that the sill corresponds to the field variance. For that reason we only provide a single sill for a directional model and this probably wont change in the future.

Cheers, Sebastian

**Reply by chibueze-oguejiofor** [reply]
Hi Sebastian

I have read a little further on this and now understand that the limitation you stated above is closely related to the concept of modeling zonal anistotropy which is usually approached by nesting two models.
I assume that GStools doesn't provide the option for nesting two models, correct?
If so, then i'd definitely love to discuss further possibilities of implementation with the community.

**Reply by LSchueler** [reply]
That sounds awesome! If you have any concrete ideas or concepts, I suggest you open a new issue where we can discuss everything, as issues are better tied to pull requests.
I'm looking forward to your ideas!

**Reply by chibueze-oguejiofor** [reply]
Great then.

Thanks again for the great work.

**Reply by MuellerSeb** [reply]
We would love to have a mechanism to construct sums of variogram models. At the moment you could do something like this:
```python
import numpy as np
import gstools as gs


class multi_cov(gs.CovModel):
    def variogram(self, r):
        r = np.array(np.abs(r), dtype=np.double)
        # exponential variogram with main var, len_scale, nugget
        result = self.var * (1 - np.exp(-r / self.len_scale)) + self.nugget
        # 1. additional exp. variogram with parameters from optional arguments
        result += self.var2 * (1 - np.exp(-r / self.len2)) + self.nug2
        # 2. additional exp. variogram with parameters from optional arguments
        result += self.var3 * (1 - np.exp(-r / self.len3)) + self.nug3
        return result


# parameter set 1
var1, len1, nug1 = 1, 2, 0.5
# parameter set 2
var2, len2, nug2 = 2, 10, 0
# parameter set 3
var3, len3, nug3 = 0.5, 20, 0

# our multi covariance model
model = multi_cov(
    var=var1,
    len_scale=len1,
    nugget=nug1,
    var2=var2,
    len2=len2,
    nug2=nug2,
    var3=var3,
    len3=len3,
    nug3=nug3,
)
# the single models for comparison
model1 = gs.Exponential(var=var1, len_scale=len1, nugget=nug1)
model2 = gs.Exponential(var=var2, len_scale=len2, nugget=nug2)
model3 = gs.Exponential(var=var3, len_scale=len3, nugget=nug3)
# plotting the different variograms
ax = model.plot(x_max=30)
model1.plot(x_max=30, ax=ax)
model2.plot(x_max=30, ax=ax)
model3.plot(x_max=30, ax=ax)
```
![Figure_1](https://user-images.githubusercontent.com/19690642/121221958-62946f80-c886-11eb-8b14-ed57cf39998e.png)

But this approach has a lot of flaws to be honest. Next problem is, that fitting such a model is almost impossible because it is over-parametrized.

**Reply by chibueze-oguejiofor** [reply]
Hi Sebastian,

Thank you so much for this. 
And yes, i do see how i would run into problems trying to fit the above model.
My understanding is that you've created three isotropic models and summed them up, such that the combined model has no anisotropy (i.e same variograms in all directions). 
I'm however trying to sum up two models (quick modification of your code above) representing directional variograms (in the main and transverse directions), SUCH THAT:  the combined model will have variograms which are different for the two different directions (represented by the two models), hence preserving anisotropy.

This way, realizations of random fields generated with the combined model will be anisotropic. Does this make sense to you?

**Reply by MuellerSeb** [reply]
That is of course true. But for now we need the assumption of ergodicity. Otherwise the `SRF` class wont be able to generate fields with the randomization method. For Kriging on the other hand, this could be interesting.

**Reply by chibueze-oguejiofor** [reply]
I see. Interesting!

Thanks again Sebastian.
